software/bismark/align/main.nf

```diff
@@ -0,0 +1,65 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process BISMARK_ALIGN {
+    tag "$meta.id"
+    label 'process_high'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:meta.id) }
+
+    conda (params.enable_conda ? "bioconda::bismark==0.23.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/bismark:0.23.0--0"
+    } else {
+        container "quay.io/biocontainers/bismark:0.23.0--0"
+    }
+
+    input:
+    tuple val(meta), path(reads)
+    path index
+
+    output:
+    tuple val(meta), path("*bam")       , emit: bam
+    tuple val(meta), path("*report.txt"), emit: report
+    tuple val(meta), path("*fq.gz")     , optional:true, emit: unmapped
+    path "*.version.txt"                , emit: version
+
+    script:
+    // Try to assign sensible bismark memory units according to what the task was given
+    def ccore = 1
+    def cpu_per_multicore = 3
+    def mem_per_multicore = (13.GB).toBytes()
+
+    if( task.cpus ){
+        // How many multicore splits can we afford with the cpus we have?
+        ccore = ((task.cpus as int) / cpu_per_multicore) as int
+        // Check that we have enough memory, assuming 13GB memory per instance (typical for mouse alignment)
+        try {
+            def tmem = (task.memory as nextflow.util.MemoryUnit).toBytes()
+            def mcore = (tmem / mem_per_multicore) as int
+            ccore = Math.min(ccore, mcore)
+        } catch (all) {
+            log.warn "Not able to define bismark align multicore based on available memory"
+        }
+
+    }
```

Comment:
This stuff still scares me 🙈

Reply:
Took this from the current implementation in nf-core/methylseq
This is a good default for mouse data I suppose, but maybe it would be better to do this sort of resource fine-tuning elsewhere?
@ewels what do you think?

Reply:
I removed the whole thing, it's probably not a good idea to just use `task.cpus` here, because it somewhat depends on available memory.
So I suggest we leave that to the pipeline code and pass it via `options.args`

Reply:
I am many (many) months too late, but I'm not sure that it is a good idea to remove this. Without it we basically can't use multi-threading for bismark alignment which drastically slows execution.

I don't think that we can (easily) set this via a config file as it has to be dynamically calculated based on the resources supplied. I wrote this code with a lot of trial and error and discussion with Felix about what Bismark is doing under the hood. It is not simple, but the code is well tested.

Although the defaults are set for mouse, I think that this is acceptable as most people will be running on Human or Mouse, or something smaller. If not then they can customise the values used to optimise.

By removing the code we just flatten to single core execution unless the user goes out of their way to customise the config to set this, which I suspect basically no-one will ever do.